airflow/jobs/job.py

```diff
@@ -183,30 +184,32 @@ def heartbeat(
         previous_heartbeat = self.latest_heartbeat
 
         try:
-            # This will cause it to load from the db
-            self._merge_from(Job._fetch_from_db(self, session))
-            previous_heartbeat = self.latest_heartbeat
-
-            if self.state == JobState.RESTARTING:
-                self.kill()
-
-            # Figure out how long to sleep for
-            sleep_for = 0
-            if self.latest_heartbeat:
-                seconds_remaining = (
-                    self.heartrate - (timezone.utcnow() - self.latest_heartbeat).total_seconds()
-                )
-                sleep_for = max(0, seconds_remaining)
-            sleep(sleep_for)
-
-            job = Job._update_heartbeat(job=self, session=session)
-            self._merge_from(job)
-
-            # At this point, the DB has updated.
-            previous_heartbeat = self.latest_heartbeat
-
-            heartbeat_callback(session)
-            self.log.debug("[heartbeat]")
+            for attempt in run_with_db_retries(logger=self.log):
```

Comment:
A retry_db_transaction decorator does exist and is used in other places where suitable

This implementation uses run_with_db_retries which is consistent with scheduler_job_runner.py and other areas.

Reply:
Yep. The decorator is used when we can retry the whole method, but this run_with_db_retries is nicer pattern when you only want to retry a block (you could also extract method and decorate the new method but in a number of cases it is not necessary and adds performance overhead. There is a delicate balance between having many methods and calling them and longer methods for readability and in this case extracting a method would be - I think a bit too much.

Reply:
That often leads to some nasty patterns like `_do_heartbeat` mehods introduced just to get decorated methods

Reply:
> These retries are required for every DB call. Instead of every where, may be its good idea to implement a wrapper that does implicitly.

BTW. If I understood what you propose, that's a bad idea to do it for every db call. This is what people get wrongly. You SHOULD NOT retry every single DB call when it fails, because in many cases it might lead to inconsistencies - because your in-memory state might get wrong. This would only be a good idea, if you have pure DB state, but if your state is shared between the in-memory objects and the DB, retrying every single DB call on operational error is wrong. By default it should fail the whole proces and restart it, to make sure that the error did not cause any in-memory vs. db inclisnsistencies. Only in cases where you are sure that you can safely retry an operation you should do it.